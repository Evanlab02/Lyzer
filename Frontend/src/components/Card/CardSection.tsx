import { PropsWithChildren } from "react";

export interface CardSectionProps extends PropsWithChildren {
    title: string;
    subtitle: string;

}

export default function CardSection(props: Readonly<CardSectionProps>) {
	const { title, subtitle, children } = props;

	return (
		<div className="lyzer-card-section">
			<h3 className="lyzer-card-section-title">{title}</h3>
			<p className="lyzer-card-section-subtitle">{subtitle}</p>
			{children}
			<hr className="lyzer-card-section-divider" />
		</div>
	);
}