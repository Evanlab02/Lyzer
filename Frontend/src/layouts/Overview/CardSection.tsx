import { DateTime } from "luxon";
import Progress from "../../components/Progress";
import Card, {CardBody, CardHeading} from "../../components/Card";
import { GridItem } from "../../components/Grid";
import { RaceWeekendProgress, UpcomingRaceWeekend } from "../../clients/interfaces/overviewInterfaces";
import { LOADING_RACE_WEEKEND_PROGRESS, LOADING_UPCOMING_RACE_WEEKEND } from "../../constants/loading";
import { convertMinutesToHighestDenominator } from "../../utils/time";


export interface CardSectionProps {
    raceWeekendProgress?: RaceWeekendProgress;
	upcomingRaceWeekend?: UpcomingRaceWeekend;
}

export default function CardSection(props: Readonly<CardSectionProps>) {
	const { 
		raceWeekendProgress = LOADING_RACE_WEEKEND_PROGRESS,
		upcomingRaceWeekend = LOADING_UPCOMING_RACE_WEEKEND
	} = props;

	return (
		<>
			<GridItem xs={12} sm={12} md={6} lg={4} xl={3} xxl={3}>
				<Card>
					<CardHeading>
                        Is it race weekend?
					</CardHeading>
					<CardBody>
						{upcomingRaceWeekend.status}
						{upcomingRaceWeekend.timeToRaceWeekend > 0 && !upcomingRaceWeekend.isRaceWeekend && (
							<p className="text-sm m-0">
								{`${convertMinutesToHighestDenominator(upcomingRaceWeekend.timeToRaceWeekend)} to go`}
							</p>
						)}
					</CardBody>
					<Progress 
						value={upcomingRaceWeekend.timeToRaceWeekendProgress}
						max={100}
						useProgressColour
					/>
				</Card>
			</GridItem>
			<GridItem xs={12} sm={12} md={6} lg={4} xl={3} xxl={3}>
				<Card>
					<CardHeading>
                        Race weekend progress
					</CardHeading>
					<CardBody>
						{raceWeekendProgress.ongoing ? `Ongoing session: ${raceWeekendProgress.name}` : `Next session: ${raceWeekendProgress.name}`}
						{raceWeekendProgress.startDateTime && (
							<p className="text-sm m-0">
								{DateTime.fromISO(raceWeekendProgress.startDateTime).toLocaleString(DateTime.DATETIME_MED_WITH_WEEKDAY)}
							</p>
						)}
					</CardBody>
					<Progress 
						value={raceWeekendProgress.weekendProgress}
						max={100}
						useProgressColour
					/>
				</Card>
			</GridItem>
			<GridItem xs={12} sm={12} md={6} lg={4} xl={3} xxl={3}>
				<Card>
					<CardHeading>
                        Driver leader
					</CardHeading>
				</Card>
			</GridItem>
			<GridItem xs={12} sm={12} md={6} lg={4} xl={3} xxl={3}>
				<Card>
					<CardHeading>
                        Constructor leader
					</CardHeading>
				</Card>
			</GridItem>
		</>
	);
}