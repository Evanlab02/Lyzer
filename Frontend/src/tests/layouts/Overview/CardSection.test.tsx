import { expect, it } from "vitest";
import { render } from "@testing-library/react";
import CardSection from "../../../layouts/Overview/CardSection";


it("CardSection renders loading state correctly", async () => {
	const { findByTestId } = render(
		<div data-testid="lyzer-card-section">
			<CardSection 
				raceWeekendProgress={undefined}
				upcomingRaceWeekend={undefined}
			/>
		</div>
	);

	const cardSection = await findByTestId("lyzer-card-section");
	expect(cardSection).toMatchSnapshot();
});

it("CardSection renders correctly", async () => {
	const { findByTestId } = render(
		<div data-testid="lyzer-card-section">
			<CardSection 
				raceWeekendProgress={{
					name: "Race",
					ongoing: true,
					weekendProgress: 100,
					startDateTime: "2025-02-22T12:00:00.000Z"
				}}
				upcomingRaceWeekend={{
					isRaceWeekend: true,
					status: "Yes",
					timeToRaceWeekendProgress: 100,
					timeToRaceWeekend: 0
				}}
			/>
		</div>
	);

	const cardSection = await findByTestId("lyzer-card-section");
	expect(cardSection).toMatchSnapshot();
});
