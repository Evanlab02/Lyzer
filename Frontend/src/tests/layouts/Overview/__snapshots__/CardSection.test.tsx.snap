// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CardSection renders correctly 1`] = `
<div
  data-testid="lyzer-card-section"
>
  <div
    class="flex-grid-item flex-grid-col-xs-12 flex-grid-col-sm-12 flex-grid-col-md-6 flex-grid-col-lg-4 flex-grid-col-xl-3 flex-grid-col-xxl-3"
  >
    <div
      class="lyzer-card"
    >
      <span
        class="lyzer-card-heading"
      >
        Is it race weekend?
      </span>
      <h3
        class="lyzer-card-body"
      >
        Yes
      </h3>
      <progress
        class="lyzer-progress lyzer-progress-complete"
        max="100"
        value="100"
      />
    </div>
  </div>
  <div
    class="flex-grid-item flex-grid-col-xs-12 flex-grid-col-sm-12 flex-grid-col-md-6 flex-grid-col-lg-4 flex-grid-col-xl-3 flex-grid-col-xxl-3"
  >
    <div
      class="lyzer-card"
    >
      <span
        class="lyzer-card-heading"
      >
        Race weekend progress
      </span>
      <h3
        class="lyzer-card-body"
      >
        Ongoing session: Race
        <p
          class="text-sm m-0"
        >
          Sat, Feb 22, 2025, 12:00 PM
        </p>
      </h3>
      <progress
        class="lyzer-progress lyzer-progress-complete"
        max="100"
        value="100"
      />
    </div>
  </div>
  <div
    class="flex-grid-item flex-grid-col-xs-12 flex-grid-col-sm-12 flex-grid-col-md-6 flex-grid-col-lg-4 flex-grid-col-xl-3 flex-grid-col-xxl-3"
  >
    <div
      class="lyzer-card"
    >
      <span
        class="lyzer-card-heading"
      >
        Driver leader
      </span>
    </div>
  </div>
  <div
    class="flex-grid-item flex-grid-col-xs-12 flex-grid-col-sm-12 flex-grid-col-md-6 flex-grid-col-lg-4 flex-grid-col-xl-3 flex-grid-col-xxl-3"
  >
    <div
      class="lyzer-card"
    >
      <span
        class="lyzer-card-heading"
      >
        Constructor leader
      </span>
    </div>
  </div>
</div>
`;

exports[`CardSection renders loading state correctly 1`] = `
<div
  data-testid="lyzer-card-section"
>
  <div
    class="flex-grid-item flex-grid-col-xs-12 flex-grid-col-sm-12 flex-grid-col-md-6 flex-grid-col-lg-4 flex-grid-col-xl-3 flex-grid-col-xxl-3"
  >
    <div
      class="lyzer-card"
    >
      <span
        class="lyzer-card-heading"
      >
        Is it race weekend?
      </span>
      <h3
        class="lyzer-card-body"
      >
        No
      </h3>
      <progress
        class="lyzer-progress lyzer-progress-incomplete"
        max="100"
        value="0"
      />
    </div>
  </div>
  <div
    class="flex-grid-item flex-grid-col-xs-12 flex-grid-col-sm-12 flex-grid-col-md-6 flex-grid-col-lg-4 flex-grid-col-xl-3 flex-grid-col-xxl-3"
  >
    <div
      class="lyzer-card"
    >
      <span
        class="lyzer-card-heading"
      >
        Race weekend progress
      </span>
      <h3
        class="lyzer-card-body"
      >
        Next session: Loading...
      </h3>
      <progress
        class="lyzer-progress lyzer-progress-incomplete"
        max="100"
        value="0"
      />
    </div>
  </div>
  <div
    class="flex-grid-item flex-grid-col-xs-12 flex-grid-col-sm-12 flex-grid-col-md-6 flex-grid-col-lg-4 flex-grid-col-xl-3 flex-grid-col-xxl-3"
  >
    <div
      class="lyzer-card"
    >
      <span
        class="lyzer-card-heading"
      >
        Driver leader
      </span>
    </div>
  </div>
  <div
    class="flex-grid-item flex-grid-col-xs-12 flex-grid-col-sm-12 flex-grid-col-md-6 flex-grid-col-lg-4 flex-grid-col-xl-3 flex-grid-col-xxl-3"
  >
    <div
      class="lyzer-card"
    >
      <span
        class="lyzer-card-heading"
      >
        Constructor leader
      </span>
    </div>
  </div>
</div>
`;
